# Pandas


Pandas is a library built using NumPy specifically for data analysis. You'll be using Pandas heavily for data manipulation, visualisation, building machine learning models, etc. 

There are two main data structures in Pandas - Series and Dataframes. The default way to store data is dataframes, and thus manipulating dataframes quickly is probably the most important skill set for data analysis. 

*Source: https://pandas.pydata.org/pandas-docs/stable/overview.html*

In this section, you will study:
1. The pandas Series (similar to a numpy array)
    * Creating a pandas series
    * Indexing series
2. Dataframes 
    * Creating dataframes from dictionaries
    * Importing CSV data files as pandas dataframes
    * Reading and summarising dataframes
    * Sorting dataframes 

### 1. The Pandas Series 

A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). 
A dataframe is simply a table where each column is a pandas series.


#### Creating Pandas Series

Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using ```pd.Series()```.

In [1]:
# import pandas, pd is an alias
import pandas as pd
import numpy as np

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
print(s)
print(type(s))

0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>


Note that each element in the Series has an index, and the index starts at 0 as usual.

In [2]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [3]:
# creating a series of type datetime
date_series = pd.date_range(start = '11-09-2017', end = '12-12-2017')
date_series
type(date_series)


pandas.core.indexes.datetimes.DatetimeIndex

#### Indexing Series

Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

In [4]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
s[3]

# accessing elements starting index = 2 till the end
s[2:]

2    5
3    6
4    9
dtype: int64

In [5]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

#### Explicitly specifying indices

You might have noticed that while creating a series, Pandas automatically indexes it from 0 to (n-1), n being the number of rows. But if we want, we can also explicitly set the index ourselves, using the ‘index’ argument while creating the series using `pd.Series()`

In [6]:
# Indexing explicitly
pd.Series([0, 1, 2], index = ['a', 'b', 'c'])

a    0
b    1
c    2
dtype: int64

In [7]:
# You can also give the index as a sequence or use functions to specify the index
# But always make sure that the number of elements in the index list is equal to the number of elements specified in the series

pd.Series(np.array(range(0,10))**2, index = range(0,10))

0     0
1     1
2     4
3     9
4    16
5    25
6    36
7    49
8    64
9    81
dtype: int32

In [20]:
import numpy as np
import pandas as pd
list_1 = [6,7,8,9,2,3,4,5] 

series_1 = pd.Series(list_1)
series_2 = series_1.apply(lambda x:x**2)

print(series_1)
print(series_2)


0    6
1    7
2    8
3    9
4    2
5    3
6    4
7    5
dtype: int64
0    36
1    49
2    64
3    81
4     4
5     9
6    16
7    25
dtype: int64


In [27]:
n = int(input())
array_1 = np.arange(1,n+1)
nn = pd.Series(data=array_1, index=array_1)
nn_squares = nn.apply(lambda x:x**2)
print(nn_squares)


4
1     1
2     4
3     9
4    16
dtype: int64


Usually, you will work with Series only as a part of dataframes. Let's study the basics of dataframes.

### The Pandas Dataframe

Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

#### Creating dataframes from dictionaries

There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc. 

In [8]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

name  age    occupation
0   Vinay   22      engineer
1  Kushal   25        doctor
2    Aman   24  data analyst
3    Saif   28       teacher

#### Importing CSV data files as pandas dataframes 

For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 



In [9]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("global_sales_data/market_fact.csv")

Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [10]:
# Looking at the top and bottom entries of dataframes
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [11]:
market_df.tail()

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [12]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
Ord_id                 8399 non-null object
Prod_id                8399 non-null object
Ship_id                8399 non-null object
Cust_id                8399 non-null object
Sales                  8399 non-null float64
Discount               8399 non-null float64
Order_Quantity         8399 non-null int64
Profit                 8399 non-null float64
Shipping_Cost          8399 non-null float64
Product_Base_Margin    8336 non-null float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.2+ KB


In [14]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [15]:
# Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [16]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [17]:
# You can extract the values of a dataframe as a numpy array using df.values 
market_df.valuesdf = pd.read_csv('https://query.data.world/s/vBDCsoHCytUSLKkLvq851k2b8JOCkF')
print(#Type your code for describing the dataset)
print(#Type your code for printing the columns of the dataset)
print(#Type your code for printing the shape of the dataset)

array([['Ord_5446', 'Prod_16', 'SHP_7609', ..., -30.51, 3.6, 0.56],
       ['Ord_5406', 'Prod_13', 'SHP_7549', ..., 4.56, 0.93, 0.54],
       ['Ord_5446', 'Prod_4', 'SHP_7610', ..., 1148.9, 2.5, 0.59],
       ...,
       ['Ord_5388', 'Prod_6', 'SHP_7524', ..., -70.85, 5.35, 0.4],
       ['Ord_5348', 'Prod_15', 'SHP_7469', ..., 565.34, 30.0, 0.62],
       ['Ord_5459', 'Prod_6', 'SHP_7628', ..., 131.39, 4.86, 0.38]],
      dtype=object)

In [21]:
df = pd.read_csv('https://query.data.world/s/vBDCsoHCytUSLKkLvq851k2b8JOCkF')
print(df.describe())
print(df.columns)
print(df.shape)

                X           Y        FFMC         DMC          DC         ISI  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean     4.669246    4.299807   90.644681  110.872340  547.940039    9.021663   
std      2.313778    1.229900    5.520111   64.046482  248.066192    4.559477   
min      1.000000    2.000000   18.700000    1.100000    7.900000    0.000000   
25%      3.000000    4.000000   90.200000   68.600000  437.700000    6.500000   
50%      4.000000    4.000000   91.600000  108.300000  664.200000    8.400000   
75%      7.000000    5.000000   92.900000  142.400000  713.900000   10.800000   
max      9.000000    9.000000   96.200000  291.300000  860.600000   56.100000   

             temp          RH        wind        rain         area  
count  517.000000  517.000000  517.000000  517.000000   517.000000  
mean    18.889168   44.288201    4.017602    0.021663    12.847292  
std      5.806625   16.317469    1.791653    0.295959    63.655

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [28]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [29]:
# Setting index to Ord_id
market_df.set_index('Ord_id', inplace = True)
market_df.head()

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes

You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [31]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = True)

Prod_id   Ship_id   Cust_id       Sales  Discount  Order_Quantity  \
Ord_id                                                                        
Ord_1      Prod_1     SHP_1    Cust_1    261.5400      0.04               6   
Ord_10     Prod_3    SHP_13   Cust_10     80.6100      0.02              15   
Ord_100    Prod_8   SHP_138   Cust_58    121.1200      0.10               3   
Ord_1000   Prod_6  SHP_1384  Cust_373    334.7100      0.01              25   
Ord_1001   Prod_5  SHP_1385  Cust_374   1981.2600      0.07              49   
Ord_1002  Prod_15  SHP_1386  Cust_375   1756.4600      0.01              15   
Ord_1003   Prod_9  SHP_1387  Cust_376     49.0400      0.02               8   
Ord_1003  Prod_11  SHP_1388  Cust_376   2320.3500      0.00               6   
Ord_1004   Prod_3  SHP_1389  Cust_365    139.5400      0.01              26   
Ord_1004   Prod_5  SHP_1390  Cust_365    199.5200      0.08              28   
Ord_1004   Prod_7  SHP_1391  Cust_365     85.6600      0.04              39   
Ord_1005  Prod_16  SHP_1392  Cust_377     15.0000      0.05               1   
Ord_1005  Prod_13  SHP_1392  Cust_377     35.4800      0.09              14   
Ord_1006   Prod_2  SHP_1393  Cust_378   1758.4100      0.06              18   
Ord_1007   Prod_4  SHP_1394  Cust_379   2766.8180      0.03              47   
Ord_1008   Prod_5  SHP_1395  Cust_380    689.6700      0.07              38   
Ord_1009   Prod_6  SHP_1397  Cust_366    237.3600      0.07              25   
Ord_1009  Prod_17  SHP_1396  Cust_366   2219.7900      0.10              30   
Ord_101    Prod_1   SHP_139   Cust_73    300.2000      0.00               8   
Ord_101    Prod_4   SHP_139   Cust_73   1364.8025      0.07              29   
Ord_1010  Prod_11  SHP_1399  Cust_366   4051.8000      0.10              35   
Ord_1010   Prod_5  SHP_1398  Cust_366    195.2300      0.08              35   
Ord_1011   Prod_3  SHP_1400  Cust_381   3465.9700      0.08               4   
Ord_1011   Prod_8  SHP_1401  Cust_381   1307.8800      0.06              13   
Ord_1012   Prod_3  SHP_1402  Cust_382    104.5100      0.04              13   
Ord_1013   Prod_4  SHP_1403  Cust_382   1064.7865      0.10              12   
Ord_1014   Prod_6  SHP_1404  Cust_368   1817.9000      0.06              33   
Ord_1014   Prod_5  SHP_1404  Cust_368    131.0900      0.01              10   
Ord_1015   Prod_9  SHP_1405  Cust_370    124.0600      0.09              17   
Ord_1016   Prod_8  SHP_1406  Cust_383    138.0500      0.09              19   
...           ...       ...       ...         ...       ...             ...   
Ord_985    Prod_4  SHP_1359  Cust_362   1155.5835      0.04              24   
Ord_986    Prod_3  SHP_1360  Cust_363     77.5700      0.00              19   
Ord_987    Prod_6  SHP_1362  Cust_365    905.3200      0.02              24   
Ord_987    Prod_4  SHP_1363  Cust_365   1702.4990      0.08              16   
Ord_988    Prod_5  SHP_1364  Cust_360    632.5500      0.05              27   
Ord_988    Prod_5  SHP_1365  Cust_360   1997.1300      0.05              32   
Ord_988    Prod_7  SHP_1364  Cust_360     90.9600      0.02              48   
Ord_989    Prod_3  SHP_1366  Cust_360     64.1300      0.10              10   
Ord_99     Prod_8   SHP_136   Cust_72    311.4400      0.04              32   
Ord_99     Prod_5   SHP_137   Cust_72    215.6500      0.01              17   
Ord_990    Prod_6  SHP_1367  Cust_366    492.7100      0.02              24   
Ord_990   Prod_11  SHP_1368  Cust_366   1042.0600      0.04              14   
Ord_991    Prod_4  SHP_1369  Cust_367    760.2485      0.04              16   
Ord_992   Prod_11  SHP_1372  Cust_368   8817.7100      0.09              27   
Ord_992   Prod_15  SHP_1370  Cust_368   5041.4600      0.06              43   
Ord_992    Prod_6  SHP_1371  Cust_368    282.9800      0.06              45   
Ord_993    Prod_1  SHP_1374  Cust_368   2645.8000      0.00              31   
Ord_993    Prod_5  SHP_1373  Cust_368   1140.2600      0.04 

In [34]:
# Sorting by values

# Sorting in increasing order of Sales
market_df.sort_values(by='Sales', ascending=True).head()

Prod_id   Ship_id    Cust_id  Sales  Discount  Order_Quantity  \
Ord_id                                                                    
Ord_704    Prod_7   SHP_964   Cust_242   2.24      0.01               1   
Ord_149    Prod_3  SHP_7028  Cust_1712   3.20      0.09               1   
Ord_4270   Prod_7  SHP_5959  Cust_1450   3.23      0.06               2   
Ord_4755  Prod_13  SHP_6628  Cust_1579   3.41      0.06               1   
Ord_2252   Prod_3  SHP_3064   Cust_881   3.42      0.05               1   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_704    -1.97           0.70                 0.37  
Ord_149    -3.16           1.49                 0.37  
Ord_4270   -2.73           0.70                 0.81  
Ord_4755   -1.78           0.70                 0.56  
Ord_2252   -2.91           1.49                 0.37

In [35]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Prod_id   Ship_id   Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361  Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561  Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55

In [38]:
# Sorting by more than two columns

# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Sales','Prod_id'], ascending = False)

Prod_id   Ship_id    Cust_id     Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_3084  Prod_17  SHP_4279  Cust_1151  89061.05      0.00              13   
Ord_2338  Prod_17  SHP_3207   Cust_932  45923.76      0.07               7   
Ord_3875  Prod_17  SHP_5370  Cust_1351  41343.21      0.09               8   
Ord_2373  Prod_14  SHP_3259   Cust_942  33367.85      0.01               9   
Ord_4614  Prod_14  SHP_6423  Cust_1571  29884.60      0.05              49   
Ord_3170  Prod_10  SHP_4400  Cust_1162  29345.27      0.03              34   
Ord_3727  Prod_17  SHP_5171  Cust_1310  29186.49      0.05              38   
Ord_997   Prod_14  SHP_1379   Cust_365  28761.52      0.04               8   
Ord_3143  Prod_14  SHP_4362  Cust_1170  28664.52      0.09              50   
Ord_3359  Prod_10  SHP_7245  Cust_1762  28389.14      0.07              33   
Ord_3707  Prod_17  SHP_5136  Cust_1307  28359.40      0.05              49   
Ord_911   Prod_10  SHP_1255   Cust_302  28180.08      0.02              32   
Ord_546   Prod_14   SHP_858   Cust_198  27875.54      0.00              46   
Ord_2425  Prod_14  SHP_3329   Cust_934  27820.34      0.08              48   
Ord_5425  Prod_14  SHP_7580  Cust_1799  27720.98      0.07              46   
Ord_825   Prod_14  SHP_1132   Cust_247  27663.92      0.05               8   
Ord_1963  Prod_11  SHP_2687   Cust_732  26622.55      0.08              49   
Ord_138   Prod_10   SHP_185    Cust_92  26133.39      0.04              30   
Ord_4216  Prod_14  SHP_5881  Cust_1432  26126.92      0.04              42   
Ord_5186  Prod_17  SHP_7247  Cust_1763  26095.13      0.03              35   
Ord_2345   Prod_3  SHP_3218   Cust_937  25409.63      0.02              20   
Ord_4161  Prod_17  SHP_5798  Cust_1421  25313.34      0.05              35   
Ord_1978  Prod_17  SHP_2703   Cust_725  25312.00      0.01              48   
Ord_2130  Prod_15  SHP_2906   Cust_796  24701.12      0.04              48   
Ord_5114  Prod_15  SHP_7144  Cust_1729  24639.80      0.01              46   
Ord_3627  Prod_17  SHP_5022  Cust_1281  24559.91      0.01              47   
Ord_1980  Prod_10  SHP_2706   Cust_924  24391.16      0.04              28   
Ord_5391  Prod_14  SHP_7527  Cust_1799  24233.54      0.07              43   
Ord_5250  Prod_17  SHP_7331  Cust_1769  24105.87      0.07              14   
Ord_1265  Prod_14  SHP_1749   Cust_466  24051.49      0.07              41   
...           ...       ...        ...       ...       ...             ...   
Ord_4490  Prod_13  SHP_6250  Cust_1507      7.96      0.00               1   
Ord_2505  Prod_13  SHP_3433   Cust_946      7.75      0.03               2   
Ord_3667   Prod_6  SHP_5083  Cust_1291      7.64      0.10               1   
Ord_3905   Prod_7  SHP_5420  Cust_1341      7.56      0.10               4   
Ord_3383   Prod_5  SHP_4690  Cust_1195      7.15      0.03               1   
Ord_4484  Prod_13  SHP_6244  Cust_1504      7.01      0.03               2   
Ord_5344  Prod_16  SHP_7462  Cust_1793      6.93      0.01               2   
Ord_3072  Prod_13  SHP_4266  Cust_1140      6.77      0.06               1   
Ord_4499  Prod_13  SHP_6265  Cust_1474      6.76      0.02               3   
Ord_2681  Prod_13  SHP_3670   Cust_992      6.75      0.08               1   
Ord_2309  Prod_13  SHP_3165   Cust_865      6.34      0.08               2   
Ord_4122  Prod_12  SHP_5743  Cust_1408      6.13      0.08               2   
Ord_3757   Prod_6  SHP_5215  Cust_1319      5.73      0.01               1   
Ord_1274   Prod_7  SHP_1760   Cust_471      5.70      0.10               1   
Ord_1218   Prod_6  SHP_1682   Cust_468      5.68      0.02               1   
Ord_3766   Prod_7  SHP_5224  Cust_1320      5.63      0.03               1   
Ord_3056   Prod_7  SHP_4245  Cust_1142      5.31      0.00               4   
Ord_834   Prod_16  SHP_1142   Cust_269      5.06      0.01               1   
Ord_104   Prod_16   SHP_142    C

In [40]:
df = pd.read_csv('https://query.data.world/s/vBDCsoHCytUSLKkLvq851k2b8JOCkF')


In [41]:
df.columns

Index(['X', 'Y', 'month', 'day', 'FFMC', 'DMC', 'DC', 'ISI', 'temp', 'RH',
       'wind', 'rain', 'area'],
      dtype='object')

In [42]:
df.head()

X  Y month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  area
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0

In [44]:
df_2 = df.set_index('X')
print(df_2.head())

   Y month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  area
X                                                                  
7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0
7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0
7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0
8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0
8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0


In [45]:
# Sort the dataframe on 'month' and 'day' in ascending order in the dataframe 'df'.

df2=df.sort_values(['month', 'day'], ascending=True)
print(df2.head(20))

     X  Y month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area
241  4  4   apr  fri  83.0   23.3   85.3   2.3  16.7  20   3.1   0.0   0.00
442  6  5   apr  mon  87.9   24.9   41.6   3.7  10.9  64   3.1   0.0   3.35
19   6  4   apr  sat  86.3   27.4   97.1   5.1   9.3  44   4.5   0.0   0.00
239  7  5   apr  sun  81.9    3.0    7.9   3.5  13.4  75   1.8   0.0   0.00
469  6  3   apr  sun  91.0   14.6   25.6  12.3  13.7  33   9.4   0.0  61.13
470  5  4   apr  sun  91.0   14.6   25.6  12.3  17.6  27   5.8   0.0   0.00
176  6  5   apr  thu  81.5    9.1   55.2   2.7   5.8  54   5.8   0.0   4.61
196  6  5   apr  thu  81.5    9.1   55.2   2.7   5.8  54   5.8   0.0  10.93
240  6  3   apr  wed  88.0   17.2   43.5   3.8  15.2  51   2.7   0.0   0.00
12   6  5   aug  fri  63.5   70.8  665.3   0.8  17.0  72   6.7   0.0   0.00
78   1  2   aug  fri  90.1  108.0  529.8  12.5  14.7  66   2.7   0.0   0.00
142  8  6   aug  fri  90.1  108.0  529.8  12.5  21.2  51   8.9   0.0   0.61
184  8  6   